File: learn/tree/binary-search-tree/binary_search_tree.go


type BSTNode struct {
	data  int
	left  *BSTNode
	right *BSTNode
}

type BinarySearchTree struct {
	root *BSTNode
}

func (bst *BinarySearchTree) Insert(data int) {
	if bst.root == nil {
		bst.root = &BSTNode{data: data}
	} else {
		bst.insertNode(bst.root, data)
	}
}

func (bst *BinarySearchTree) insertNode(node *BSTNode, data int) {
	if data < node.data {
		if node.left == nil {
			node.left = &BSTNode{data: data}
		} else {
			bst.insertNode(node.left, data)
		}
	} else {
		if node.right == nil {
			node.right = &BSTNode{data: data}
		} else {
			bst.insertNode(node.right, data)
		}
	}
}

func (bst *BinarySearchTree) Search(data int) bool {
	return bst.searchNode(bst.root, data)
}

func (bst *BinarySearchTree) searchNode(node *BSTNode, data int) bool {
	if node == nil {
		return false
	}
	if data < node.data {
		return bst.searchNode(node.left, data)
	} else if data > node.data {
		return bst.searchNode(node.right, data)
	}
	return true
}

File: learn/tree/binary-search-tree/binary_search_tree.rs
pub struct BSTNode {
    pub data: i32,
    pub left: Option<Box<BSTNode>>,
    pub right: Option<Box<BSTNode>>,
}

pub struct BinarySearchTree {
    pub root: Option<Box<BSTNode>>,
}

impl BinarySearchTree {
    pub fn new() -> Self {
        BinarySearchTree { root: None }
    }

    pub fn insert(&mut self, data: i32) {
        if let Some(ref mut node) = self.root {
            Self::insert_node(node, data);
        } else {
            self.root = Some(Box::new(BSTNode {
                data,
                left: None,
                right: None,
            }));
        }
    }

    fn insert_node(node: &mut Box<BSTNode>, data: i32) {
        if data < node.data {
            if let Some(ref mut left) = node.left {
                Self::insert_node(left, data);
            } else {
                node.left = Some(Box::new(BSTNode {
                    data,
                    left: None,
                    right: None,
                }));
            }
        } else {
            if let Some(ref mut right) = node.right {
                Self::insert_node(right, data);
            } else {
                node.right = Some(Box::new(BSTNode {
                    data,
                    left: None,
                    right: None,
                }));
            }
        }
    }

    pub fn search(&self, data: i32) -> bool {
        Self::search_node(&self.root, data)
    }

    fn search_node(node: &Option<Box<BSTNode>>, data: i32) -> bool {
        if let Some(ref n) = node {
            if data < n.data {
                return Self::search_node(&n.left, data);
            } else if data > n.data {
                return Self::search_node(&n.right, data);
            } else {
                return true;
            }
        }
        false
    }
}
